hyperspy/learn/rpca.py

```diff
@@ -0,0 +1,189 @@
+    ----------
+    X : numpy array
+        is the [m x n] matrix of observations.
+    rank : int
+        The model dimensionality.
+    lambda1 : float
+        Nuclear norm regularization parameter.
+    lambda2 : float
+        Sparse error regularization parameter.
+    method : 'CF' | 'BCD'
+        CF  - Closed-form (default)
+        BCD - Block-coordinate descent
+    mask : numpy array
+        is an initial estimate of the sparse error matrix
+    seed : int
+        Random seed for initialization
+
+    Returns
+    -------
+    L : numpy array
+        is the [m x r] basis array
+    R : numpy array
+        is the [r x n] coefficient array
+    E : numpy array
+        is the sparse error
+    U, S, V : numpy arrays
+        are the results of an SVD on Y
+
+    """
+    if fast is True and sklearn_installed is True:
+        def svd(X):
+            return fast_svd(X, rank)
+    else:
+        def svd(X):
+            return scipy.linalg.svd(X, full_matrices=False)
+
+    # Get shape
+    m, n = X.shape
+
+    # Check options if None
+    if method is None:
+        _logger.warning("No method specified. Defaulting to "
+                        "closed-form solver")
+        method = 'CF'
+    if lambda1 is None:
+        _logger.warning("Nuclear norm regularization parameter "
+                        "is set to default.")
+        lambda1 = 1.0 / np.sqrt(n)
+    if lambda2 is None:
+        _logger.warning("Sparse regularization parameter "
+                        "is set to default.")
+        lambda2 = 1.0 / np.sqrt(n)
+
+    # Check options are valid
+    if method not in ('CF', 'BCD'):
+        raise ValueError("'method' not recognised")
+
+    # Set random seed
+    if seed is not None:
+        np.random.seed(seed)
+
+    # Use random initialization
+    Y2 = np.random.randn(m, rank)
+    L, tmp = scipy.linalg.qr(Y2, mode='economic')
+
+    R = np.zeros((rank, n))
+    I = lambda1 * np.eye(rank)
+
+    # Allow the error matrix to be initialized
+    if mask is None:
+        E = np.zeros((m, n))
+    else:
+        E = mask.reshape((m, n))
+
+    A = np.zeros((rank, rank))
+    B = np.zeros((m, rank))
+
+    for t in range(n):
+        if t == 0 or np.mod(t + 1, np.round(n / 10)) == 0:
+            _logger.info("Processing sample : %s" % (t + 1))
+
+        z = X[:, t]
+        r, e = _solveproj(z, L, I, lambda2)
+
+        R[:, t] = r
+        E[:, t] = e
+
+        if method == 'CF':
+            # Closed-form solution
+            A = A + np.outer(r, r.T)
+            B = B + np.outer((z - e), r.T)
+            L = np.dot(B, scipy.linalg.inv(A + I))
+        else:
+            # Block-coordinate descent
+            A = A + np.outer(r, r.T)
+            B = B + np.outer((z - e), r.T)
+            L = _updatecol(L, A, B, I)
+
+    # Do final SVD
+    U, S, Vh = svd(np.dot(L, R))
+    V = Vh.T
+
+    # Chop small singular values which
+    # likely arise from numerical noise
+    # in the SVD.
+    S[S<=1e-9] = 0.0
```

Comment:
This could be better achieved by ensuring all singular values above `rank` are set to zero. e.g. `S[rank:] = 0.`.